org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/UriExtensions.xtend

```diff
@@ -21,32 +27,100 @@ import java.nio.charset.StandardCharsets
 @Singleton
 class UriExtensions {
 
+	static val ESCAPED_FILE_SCHEME = URLEncoder.encode("file:", StandardCharsets.UTF_8.name);
+
+	/**
+	 * Converts any non-{@code null} arguments, representing a URI as a string, into a EMF based URI where the following
+	 * rules apply:
+	 * <p>
+	 * <ul>
+	 * <li>If the argument has a {@code file} scheme:
+	 * <ul>
+	 * <li>and the {@code host} is not set (e.g.: {@code localhost} can be omitted) then it returns with a EMF URI with the
+	 * following form: {@code file:///path/to/resource}.</li>
+	 * <li>and the {@code host} is set (e.g.: {@code localhost} can be omitted) then drops the URI host and returns with a 
+	 * EMF URI with the following form: {@code file:///path/to/resource}.</li>
+	 * <li>if the URI starts with two forward-slashes ({@code file://path/to/resource}), the method will treat it with a URI
+	 * with {@code localhost} scheme and will omit it, producing the {@code file:///path/to/resource} result.</li>
+	 * </ul>
+	 * <li>If the has any other scheme then it will return with EMF URI {@link URI#createURI(String) created} from the
+	 * argument.</li>
+	 * <li>This method takes care of the encoding and the URI path escaping.</li>
+	 * </ul>
+	 * </p>
+	 */
 	def URI toUri(String pathWithScheme) {
-		// URI is used to get path of the uri without scheme
-		var path = URI.createURI(pathWithScheme).path
-		return URI.createURI(path.toPath)
+		// This will take care of the encoding.
+		val uri = URI.createURI(pathWithScheme.checkNotNull('pathWithScheme'));
+		if (uri.file) {
+			val segments = uri.toSegments;
+			val prefix = try {
```

Comment:
I think the locations that assume a uri with a trailing empty segment (prefix-uri) should take care. 
Handling it in a central manner will alter URIs surprisingly in other cases, also it won't work for non existing directories but should.
We already do stuff like that here : https://github.com/eclipse/xtext-core/blob/master/org.eclipse.xtext/src/org/eclipse/xtext/workspace/FileProjectConfig.xtend#L98

Reply:
I couldn't find a location where the `org.eclipse.xtext.util.UriUtil.isPrefixOf(URI, URI)` is used with a uri that has not been converted. Why was it added here?
If you find such place, please use `org.eclipse.xtext.util.UriUtil.toFolderURI(URI)`.

Reply:
Maybe not `isPrefixOf` but `resolve` caused the issue. We should make sure that `FileProjectConfig` is in all cases initialized with a "FolderUri" (see URIUtils).